File: js/darts.js
"use strict";

// Port of darts.java, including the Dart helper class.

class DartProjectile {
  constructor(tx, ty, sp) {
    const side = Math.floor(Math.random() * 3);
    switch (side) {
      case 0:
        this.x = Math.random() * 400;
        this.y = 300;
        break;
      case 1:
        this.x = 0;
        this.y = 160 + Math.random() * (400 - 160);
        break;
      case 2:
        this.x = 400;
        this.y = 160 + Math.random() * (400 - 160);
        break;
    }
    this.dir = this.dir_to(this.x, this.y, tx, ty);
    this.speed = sp;
    this.used = true;
  }

  // Returns true when the dart hits the rectangle r {x,y,w,h}.
  go(g, r) {
    this.x += this.speed * Math.cos(this.dir);
    this.y -= this.speed * Math.sin(this.dir);
    const p = { x: Math.trunc(this.x), y: Math.trunc(this.y) };

    const opdir = Math.PI - this.dir;
    const halfpi = Math.PI / 2;
    const back = this.movePoint(p, opdir, 12);
    const bl = this.movePoint(back, opdir - halfpi, 2);
    const br = this.movePoint(back, opdir + halfpi, 2);
    const taa = this.movePoint2(p, opdir, 3, opdir - halfpi, 2);
    const tab = this.movePoint2(p, opdir, 3, opdir + halfpi, 2);
    const tba = this.movePoint2(p, opdir, 9, opdir - halfpi, 2);
    const tbb = this.movePoint2(p, opdir, 9, opdir + halfpi, 2);

    g.strokeStyle = "#fff";
    g.beginPath();
    g.moveTo(p.x, p.y);
    g.lineTo(back.x, back.y);
    g.moveTo(bl.x, bl.y);
    g.lineTo(br.x, br.y);
    g.stroke();

    g.fillStyle = "red";
    g.beginPath();
    g.moveTo(taa.x, taa.y);
    g.lineTo(tab.x, tab.y);
    g.lineTo(tbb.x, tbb.y);
    g.lineTo(tba.x, tba.y);
    g.closePath();
    g.fill();

    if (this.x < 0 || this.x > 400 || this.y < 120 || this.y > 300) {
      this.used = false;
    }

    const ix = Math.trunc(this.x), iy = Math.trunc(this.y);
    return ix >= r.x && ix < r.x + r.w && iy >= r.y && iy < r.y + r.h;
  }

  dir_to(x1, y1, x2, y2) {
    if (x1 === x2) {
      return (y1 > y2) ? Math.PI / 2 : Math.PI * 3 / 2;
    }
    if (y1 === y2) {
      return (x1 > x2) ? Math.PI : 0;
    }
    const yleg = y2 - y1;
    const xleg = x2 - x1;
    const hyp = Math.sqrt(yleg * yleg + xleg * xleg);
    let ref = Math.asin(yleg / hyp);
    if (x2 < x1) ref = Math.PI + ref;
    else ref = -ref;
    return ref;
  }

  movePoint(p, d1, s1) {
    return {
      x: Math.trunc(p.x + Math.cos(d1) * s1),
      y: Math.trunc(p.y + Math.sin(d1) * s1),
    };
  }

  movePoint2(p, d1, s1, d2, s2) {
    return {
      x: Math.trunc(p.x + Math.cos(d1) * s1 + Math.cos(d2) * s2),
      y: Math.trunc(p.y + Math.sin(d1) * s1 + Math.sin(d2) * s2),
    };
  }
}

class DartsLevel extends KLevel {
  static SPEED_CHANGE = 7;
  static TOP_LIMIT = 173;
  static NUMDARTS = 50;

  constructor(game) {
    super(game);
    this.plot = "Larry ist am Ostportal";
    this.inst = "Dodge the darts and last as long as you can!!";

    this.man = Array.from({ length: 3 }, () => []);
    this.lasth = null;
    this.lastv = null;
    this.dx = 200; this.dy = 200;
    this.df = 0; this.dsx = 0; this.dsy = 0;
    this.ld = 0;  // 0 - face right, 1 - face left
    this.walkto = null;
    this.dart = new Array(DartsLevel.NUMDARTS).fill(null);
    this.dart_at = 0;
    this.dart_to = 30;
  }

  async loadData() {
    this.scene = await loadImage("assets/dart_scene.gif");
    const allman = await loadImage("assets/dart_man.gif");
    for (let y = 0; y < 2; y++) {
      for (let x = 0; x < 3; x++) {
        this.man[x][y] = cropImage(allman, x * 40, y * 83, 40, 83);
      }
    }
    this.speed = 50;
    this.start();
  }

  drawStatic(g) {
    g.drawImage(this.scene, 0, 0);
  }

  drawNormal(g) {
    const S = DartsLevel.SPEED_CHANGE;

    // mouse movements, if applicable
    if (this.walkto !== null) {
      this.dsx = 0;
      this.dsy = 0;
      if (this.walkto.x < this.dx) { this.dsx = -S; this.ld = 1; }
      if (this.walkto.x > this.dx) { this.dsx = S; this.ld = 0; }
      if (this.walkto.y < this.dy) this.dsy = -S;
      if (this.walkto.y > this.dy) this.dsy = S;

      if (Math.abs(this.walkto.x - this.dx) < S) this.walkto.x = this.dx;
      if (Math.abs(this.walkto.y - this.dy) < S) this.walkto.y = this.dy;
      if (this.walkto.x === this.dx && this.walkto.y === this.dy) {
        this.walkto = null;
        this.dsx = 0;
        this.dsy = 0;
      }
    }

    // move the man
    this.dx += this.dsx;
    this.dy += this.dsy;
    if (this.dx < 0) this.dx = 0;
    if (this.dx > 400) this.dx = 400;
    if (this.dy < DartsLevel.TOP_LIMIT) this.dy = DartsLevel.TOP_LIMIT;
    if (this.dy > 341) this.dy = 341;

    // if applicable, add more darts
    if (++this.dart_at > this.dart_to) {
      this.dart_at = 0;
      let fd = -1;
      for (let i = 0; i < DartsLevel.NUMDARTS; i++) {
        if (fd === -1 && (this.dart[i] === null || this.dart[i].used === false)) {
          fd = i;
        }
      }
      if (fd !== -1) {
        this.dart[fd] = new DartProjectile(this.dx, this.dy - 40, 5);
      }
      if ((this.dart_to -= 1) < 2) this.dart_to = 2;

      this.changeScore(5);
    }

    if (++this.df > 2) this.df = 0;
    const moving = (this.dsx !== 0 || this.dsy !== 0);
    g.drawImage(this.man[moving ? this.df : 1][this.ld], this.dx - 20, this.dy - 83);

    const r = { x: this.dx - 15, y: this.dy - 80, w: 30, h: 60 };

    let dead = false;
    for (let i = 0; i < DartsLevel.NUMDARTS; i++) {
      if (this.dart[i] !== null && this.dart[i].used) {
        if (this.dart[i].go(g, r)) dead = true;
      }
    }

    if (dead) {
      this.stop();
      this.enableNextLevel();
      this.displayText(g, "Du bist betäubt...");
    }
  }

  keyDown(key) {
    const S = DartsLevel.SPEED_CHANGE;
    switch (key) {
      case "left":
        this.lasth = key;
        this.dsx = -S;
        this.ld = 1;
        break;
      case "right":
        this.lasth = key;
        this.dsx = S;
        this.ld = 0;
        break;
      case "up":
        this.lastv = key;
        this.dsy = -S;
        break;
      case "down":
        this.lastv = key;
        this.dsy = S;
        break;
    }
  }

  keyUp(key) {
    switch (key) {
      case "left":
      case "right":
        if (key === this.lasth) this.dsx = 0;
        break;
      case "up":
      case "down":
        if (key === this.lastv) this.dsy = 0;
        break;
    }
  }

  mouseDown(x, y) {
    this.lasth = null;
    this.lastv = null;
    this.mouseDrag(x, y);
  }

  mouseDrag(x, y) {
    if (y < DartsLevel.TOP_LIMIT) y = DartsLevel.TOP_LIMIT;
    this.walkto = { x, y };
  }

  mouseUp(x, y) {
    this.dsx = 0;
    this.dsy = 0;
  }
}

File: darts.java
import java.awt.*;
import java.applet.AudioClip;

public class darts extends klevel
{
    public final static int SPEED_CHANGE = 7, TOP_LIMIT = 173;

    Image[][] man = new Image[3][2];
    Image scene;
    int lasth, lastv;  // last horiz and vert key directions
    int dx=200, dy=200, df, dsx, dsy;
    int ld = 0; // 0-face right, 1-faceleft

    Point walkto;

    public final static int NUMDARTS = 50;
    Dart[] dart = new Dart[NUMDARTS];
    int dart_at = 0, dart_to = 30;

    public darts ()
    {
        plot = "Larry ist am Ostportal";
        inst = "Dodge the darts and last as long as you can!!";
    }

    public void loadData ()
    {
        scene = waitImage("dart_scene.gif");

        Image allman = waitImage("dart_man.gif");
        for (int y=0;y<2;y++)
            for (int x=0;x<3;x++)
                man[x][y] = cropImage(allman, x*40, y*83, 40, 83);

        speed=50;
        start();
    }

    public void drawNormal (Graphics g)
    {
        // mouse movements, if applicable
        if (walkto != null)
        {
            dsx=0; dsy=0;
            if (walkto.x<dx) { dsx=-SPEED_CHANGE; ld=1; }
            if (walkto.x>dx) { dsx=SPEED_CHANGE; ld=0;  }
            if (walkto.y<dy) dsy=-SPEED_CHANGE;
            if (walkto.y>dy) dsy=SPEED_CHANGE;

            if ((int)(Math.abs((double)(walkto.x-dx))) < SPEED_CHANGE) walkto.x = dx;
            if ((int)(Math.abs((double)(walkto.y-dy))) < SPEED_CHANGE) walkto.y = dy;
            if (walkto.x==dx && walkto.y==dy)
            {
                walkto=null;
                dsx = 0;
                dsy = 0;
            }
        }

        // move the man
        dx += dsx;
        dy += dsy;
        if (dx<0) dx=0;
        if (dx>400) dx=400;
        if (dy<TOP_LIMIT) dy=TOP_LIMIT;
        if (dy>341) dy=341;

        // if applicable, add more darts
        if (++dart_at>dart_to)
        {
            dart_at=0;
            int fd = -1;
            for (int i=0;i<NUMDARTS;i++)
            {
                if (fd==-1 && (dart[i]==null || dart[i].used==false)) fd = i;
            }
            if (fd != -1)
            {
                dart[fd] = new Dart(dx, dy-40, 5);
            }
            if ((dart_to-=1)<2) dart_to=2;

            changeScore(5);
        }

        if (++df>2) df=0;
        g.drawImage(man[(dsx!=0||dsy!=0)?df:1][ld],dx-20,dy-83,this);

        Rectangle r = new Rectangle(dx-15,dy-80,30,60);

        boolean dead = false;
        for (int i=0;i<NUMDARTS;i++)
            if (dart[i]!=null && dart[i].used)
                if (dart[i].go(g, r)) dead = true;

        if (dead)
        {
            stop();
            enableNextLevel();
            displayText(g,"Du bist betäubt...");
        }
    }

    public void drawStatic (Graphics g)
    {
        g.drawImage(scene,0,0,this);
    }

    public boolean keyDown(Event evt, int key)
    {
        switch (key)
        {
            case Event.LEFT:
                lasth = key;
                dsx = -SPEED_CHANGE;
                ld=1;
                break;

            case Event.RIGHT:
                lasth = key;
                dsx = SPEED_CHANGE;
                ld=0;
                break;

            case Event.UP:
                lastv = key;
                dsy = -SPEED_CHANGE;
                break;

            case Event.DOWN:
                lastv = key;
                dsy = SPEED_CHANGE;
                break;
        }
        return true;
    }

    public boolean keyUp(Event evt, int key)
    {
        switch (key)
        {
            case Event.LEFT:
                if (key == lasth) dsx = 0;
                break;

            case Event.RIGHT:
                if (key == lasth) dsx = 0;
                break;

            case Event.UP:
                if (key == lastv) dsy = 0;
                break;

            case Event.DOWN:
                if (key == lastv) dsy = 0;
                break;
        }
        return true;
    }

    public boolean mouseDown (Event e, int x, int y)
    {
        lasth = 0;
        lastv = 0;
        return mouseDrag (e, x, y);
    }

    public boolean mouseDrag (Event e, int x, int y)
    {
        if (y<TOP_LIMIT) y = TOP_LIMIT;
        walkto = new Point(x, y);
        return true;
    }

    public boolean mouseUp (Event e, int x, int y)
    {
//      walkto = null;
        dsx = 0;
        dsy = 0;
        return true;
    }

}

class Dart
{
    boolean used = false;
    double x, y, dir, speed;

    public Dart (int tx, int ty, int sp)
    {
        int side = (int)(Math.random()*3);
        switch (side)
        {
            case 0:
                x = Math.random()*400;
                y = 300;
                break;
            case 1:
                x = 0;
                y = 160d+Math.random()*(400-160);
                break;
            case 2:
                x = 400;
                y = 160d+Math.random()*(400-160);
                break;
        }
        dir = dir_to(x,y,(double)tx,(double)ty);
        speed = sp;
        used = true;
    }

    public boolean go (Graphics g, Rectangle r)
    {
        x += speed * Math.cos(dir);
        y -= speed * Math.sin(dir);
        Point p = new Point((int)x, (int)y);

        /*

            |        picture is to scale, | is white, X is red
            |
            |
           XXX
           XXX
           XXX
           XXX
           XXX
           XXX
           XXX
            |
            |
          --+--

        */

        double opdir = Math.PI-dir, halfpi = Math.PI/2;
        Point back = movePoint(p,opdir,12);
        Point bl = movePoint(back,opdir-halfpi,2);
        Point br = movePoint(back,opdir+halfpi,2);
        Point taa = movePoint(p,opdir,3,opdir-halfpi,2);
        Point tab = movePoint(p,opdir,3,opdir+halfpi,2);
        Point tba = movePoint(p,opdir,9,opdir-halfpi,2);
        Point tbb = movePoint(p,opdir,9,opdir+halfpi,2);
        int[] px = { taa.x, tab.x, tbb.x, tba.x, taa.x };
        int[] py = { taa.y, tab.y, tbb.y, tba.y, taa.y };

        g.setColor(Color.white);
        g.drawLine(p.x,p.y,back.x,back.y);
        g.drawLine(bl.x,bl.y,br.x,br.y);

        g.setColor(Color.red);
        g.fillPolygon(px, py, 5);

        if (x<0 || x>400 || y<120 || y>300) used=false;

        return r.inside((int)x, (int)y);
    }

    public double dir_to (double x1, double y1, double x2, double y2)
    {
        double ret=-1;

        if (x1==x2)
        {
            if (y1>y2)
                ret=Math.PI/2;
            else
                ret=Math.PI*3/2;
        } else
        if (y1==y2)
        {
            if (x1>x2)
                ret=Math.PI;
            else
                ret=0;
        } else
        {
            double yleg= y2 - y1;
            double xleg= x2 - x1;
            double hyp=Math.sqrt((yleg*yleg)+(xleg*xleg));

            double ref=Math.asin(yleg/hyp);

            if (x2<x1) ref=Math.PI + ref; else ref=-ref;
            ret = ref;
        }

        return ret;
    }

    public Point movePoint (Point p, double d1, double s1, double d2, double s2)
    {
        double x=(double)p.x, y=(double)p.y;

        x+=Math.cos(d1) * s1;
        y+=Math.sin(d1) * s1;
        x+=Math.cos(d2) * s2;
        y+=Math.sin(d2) * s2;
        return new Point((int)x,(int)y);
    }

    public Point movePoint (Point p, double d1, double s1)
    {
        double x=(double)p.x, y=(double)p.y;

        x+=Math.cos(d1) * s1;
        y+=Math.sin(d1) * s1;
        return new Point((int)x,(int)y);
    }

}